cpu/cc26x2_cc13x2/setup.c

```diff
@@ -0,0 +1,197 @@
+     * wakes up from sleep */
+    FLASH->FPAC1 = (FLASH->FPAC1 & ~FLASH_FPAC1_PSLEEPTDIS_m) |
+                   (0x139 << FLASH_FPAC1_PSLEEPTDIS_s);
+
+    /* Finally at the end of the flash boot process:
+     * Set BOOT_DET bits */
+    _set_boot_det();
+
+    /* Make sure there are no ongoing VIMS mode change when leaving (There
+     * should typically be no wait time here, but need to be sure)
+     */
+    while (VIMS->STAT & VIMS_STAT_MODE_CHANGING) {}
+}
+
+static void _set_lpm_bias(void)
+{
+    /* Propagate the LPM_BIAS trim */
+    uint32_t trim_reg = FCFG->DAC_BIAS_CNF;
+    uint32_t trim_value = ((trim_reg & FCFG1_DAC_BIAS_CNF_LPM_TRIM_IOUT_m) >>
+                          FCFG1_DAC_BIAS_CNF_LPM_TRIM_IOUT_s);
+    adi_write(ADI_4_AUX_BASE, ADI_4_AUX_LPMBIAS,
+              (trim_value << ADI_4_AUX_LPMBIAS_LPM_TRIM_IOUT_s) &
+              ADI_4_AUX_LPMBIAS_LPM_TRIM_IOUT_m);
+
+    /* Set LPM_BIAS_BACKUP_EN according to FCFG configuration */
+    if (trim_reg & FCFG1_DAC_BIAS_CNF_LPM_BIAS_BACKUP_EN) {
+        adi_set(ADI3_BASE, ADI_3_REFSYS_AUX_DEBUG,
+                ADI_3_REFSYS_AUX_DEBUG_LPM_BIAS_BACKUP_EN);
+    }
+    else {
+        adi_clr(ADI3_BASE, ADI_3_REFSYS_AUX_DEBUG,
+                ADI_3_REFSYS_AUX_DEBUG_LPM_BIAS_BACKUP_EN);
+    }
+
+    /* Set LPM_BIAS_WIDTH_TRIM according to FCFG1 configuration */
+    uint32_t width_trim =
+        ((trim_reg & FCFG1_DAC_BIAS_CNF_LPM_BIAS_WIDTH_TRIM_m) >>
+        FCFG1_DAC_BIAS_CNF_LPM_BIAS_WIDTH_TRIM_s);
+    width_trim = width_trim << ADI_4_AUX_COMP_LPM_BIAS_WIDTH_TRIM_s;
+    adi_write_8bits(ADI_4_AUX_BASE, ADI_4_AUX_COMP,
+                    ADI_4_AUX_COMP_LPM_BIAS_WIDTH_TRIM_m, width_trim);
+}
+
+static void _set_dcdc_settings(void)
+{
+    /* Check in CCFG for alternative DCDC setting */
+    if ((CCFG->SIZE_AND_DIS_FLAGS &
+        CCFG_SIZE_AND_DIS_FLAGS_DIS_ALT_DCDC_SETTING) == 0) {
+        /**
+         * ADI_3_REFSYS:DCDCCTL5[3] = CCFG:MODE_CONF_1[19] (=ALT_DCDC_DITHER_EN)
+         * ADI_3_REFSYS:DCDCCTL5[2:0] = CCFG:MODE_CONF_1[18:16](=ALT_DCDC_IPEAK)
+         *
+         * Using a single 4-bit masked write since layout is equal for both
+         * source and destination
+         */
+        uint32_t dcdc_setting = CCFG->MODE_CONF_1 >>
+                                CCFG_MODE_CONF_1_ALT_DCDC_IPEAK_s;
+        adi_write_4bits(ADI3_BASE, ADI_3_REFSYS_DCDCCTL5, dcdc_setting, 0xF0);
+    }
+
+    /* Force DCDC to use RCOSC before starting up XOSC.
+     * Clock loss detector does not use XOSC until SCLK_HF actually switches
+     * and thus DCDC is not protected from clock loss on XOSC in that time
+     * frame. The force must be released when the switch to XOSC has happened.
+     * This is done in OSCHfSourceSwitch().
+     */
+    reg32_t *addr = (reg32_t *)(AUX_DDI0_OSC_BASE + (DDI_0_OSC_CTL0 << 1) +
+                                DDI_MASK16B + 4);
+    *addr = DDI_0_OSC_CTL0_CLK_DCDC_SRC_SEL_m |
+            (DDI_0_OSC_CTL0_CLK_DCDC_SRC_SEL_m >> 16);
```

Comment:
Is there no `struct` for that?

Reply:
There's an struct for `DDI_0_OSC`. But TI prefers to use their "masked access instructions" to write into those registers, I suspect that's to avoid the read-write-modify cycle (DDI might modify the register and some bits will be lost I suspect).

Their way of writing is using masked access in that case using a 32-bit register to write 16-bits of data depending on the bits set on `mask`:

```
[mask][data]
  |           | uint16_t
  | uint16_t
```

There might a way to setup an struct for masked writes, however I don't know if it's worth the effort right now, it might be needed more when we need to write ADC/DAC drivers, a that would be needed for each register (or at least a common one for each instruction).

I can write a function to write to these registers as I did with `adi_write_8bits` type of functions.